packages/google_sign_in/lib/google_sign_in.dart

```diff
@@ -126,31 +119,80 @@ class GoogleSignIn {
     await _initialization;
     Map<String, dynamic> response = await _channel.invokeMethod(method);
     _currentUser = (response != null && response.isNotEmpty)
-        ? new GoogleSignInAccount._(this, response) : null;
+        ? new GoogleSignInAccount._(this, response)
+        : null;
     _streamController.add(_currentUser);
     return _currentUser;
   }
 
-  /// The currently signed in account, or null if the user is signed out
-  GoogleSignInAccount _currentUser;
+  /// Keeps track of the most recently scheduled method call.
+  _MethodCompleter _lastMethodCompleter;
+
+  /// Adds call to [method] in a queue for execution.
+  ///
+  /// At most one in flight call is allowed to prevent concurrent (out of order)
+  /// updates to [currentUser] and [onCurrentUserChanged].
+  Future<GoogleSignInAccount> _addMethodCall(String method) {
+    if (_lastMethodCompleter == null) {
+      _lastMethodCompleter = new _MethodCompleter(method);
+      _lastMethodCompleter.complete(_callMethod(method));
+      return _lastMethodCompleter.future;
+    }
+
+    // Return the same Future for consecutive calls to the same method.
+    if (_lastMethodCompleter.method == method &&
+        !_lastMethodCompleter.isCompleted) {
```

Comment:
We don't know if the last call completed successfully or with an error.
This only covers a situation when a second call to the same method comes while the first one is in-flight. Handling completed calls is done on lines 150-158.

This is actually more like a small optimization. Without this `if` block we'll get the same guarantees, the only difference is that each method call would always return a fresh Future instance.

So it should be safe to remove it, which maybe not a bad idea (less code, less troubles).